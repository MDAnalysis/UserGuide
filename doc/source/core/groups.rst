.. -*- coding: utf-8 -*-

====================
Groups
====================


For convenience, chemically meaningful groups of Atoms such as a 
Residue or a Segment (typically a whole molecule or all of the solvent) 
also exist as containers, as well as groups of these units (
ResidueGroup, SegmentGroup).

AtomGroups
==========
AtomGroup instances can be easily created 
(e.g., from an AtomGroup.select_atoms() selection or simply by slicing).

