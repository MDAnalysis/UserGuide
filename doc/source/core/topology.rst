.. -*- coding: utf-8 -*-

====================
Topology
====================

MDA Topology =/= 'force field' topology. Connectivity only. See ParmEd etc for other stuff.

Adding attributes to Topology
=============================

stuff