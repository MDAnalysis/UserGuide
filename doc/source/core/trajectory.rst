.. -*- coding: utf-8 -*-

====================
Trajectory
====================

Not intuitive.

.positions() updates

Auxiliary trajectory readers
============================

