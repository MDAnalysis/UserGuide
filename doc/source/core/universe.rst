.. -*- coding: utf-8 -*-

====================
Universes
====================

| *If you wish to make an apple pie from scratch, you must first invent the universe.*
| - Carl Sagan, Cosmos

| *In the beginning the Universe was created.*
| *This has made a lot of people very angry and been widely regarded as a bad move.*
| â€• Douglas Adams, The Restaurant at the End of the Universe

The Universe class ties a topology and a trajectory together. 
Almost all code in MDAnalysis starts with a Universe.

Creating a Universe
===================

Loading from files
------------------

.. code-block:: python

u = Universe(topology, trajectory)          # read system from file(s)
u = Universe(pdbfile)                       # read atoms and coordinates from PDB or GRO
u = Universe(topology, [traj1, traj2, ...]) # read from a list of trajectories
u = Universe(topology, traj1, traj2, ...)   # read from multiple trajectories

Many words here

Constructing from AtomGroups
----------------------------

Merge()

