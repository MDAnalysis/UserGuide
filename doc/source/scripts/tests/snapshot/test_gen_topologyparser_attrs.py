
from gen_topologyparser_attrs import TopologyParsers, TopologyAttrs, ConnectivityAttrs
from MDAnalysis.core import selection as sel
from unittest.mock import patch

def test_TopologyParsers(snapshot):
    with patch("builtins.open"):
        top = TopologyParsers()
    assert top.lines == snapshot

def test_TopologyAttrs(snapshot):
    with patch("builtins.open"):
        top = TopologyParsers()
        ta = TopologyAttrs(top.attrs)
    assert ta.lines == snapshot

def test_ConnectivityAttrs(snapshot):
    with patch("builtins.open"):
        top = TopologyParsers()
        ca = ConnectivityAttrs(top.attrs)
    assert ca.lines == snapshot
