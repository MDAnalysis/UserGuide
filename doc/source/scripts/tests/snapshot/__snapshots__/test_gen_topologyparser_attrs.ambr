# serializer version: 1
# name: test_ConnectivityAttrs
  list([
    list([
      'angles',
      'angles',
      ':ref:`DATA <DATA-format>`, :ref:`GSD <GSD-format>`, :ref:`PSF <PSF-format>`, :ref:`TOP, PRMTOP, PARM7 <TOP-format>`, :ref:`TPR <TPR-format>`, :ref:`XML <XML-format>`',
    ]),
    list([
      'bonds',
      'bonds',
      ':ref:`DATA <DATA-format>`, :ref:`DMS <DMS-format>`, :ref:`GSD <GSD-format>`, :ref:`MMTF <MMTF-format>`, :ref:`MOL2 <MOL2-format>`, :ref:`PSF <PSF-format>`, :ref:`TOP, PRMTOP, PARM7 <TOP-format>`, :ref:`TPR <TPR-format>`, :ref:`TXYZ, ARC <TXYZ-format>`, :ref:`XML <XML-format>`',
    ]),
    list([
      'dihedrals',
      'dihedrals',
      ':ref:`DATA <DATA-format>`, :ref:`GSD <GSD-format>`, :ref:`PSF <PSF-format>`, :ref:`TOP, PRMTOP, PARM7 <TOP-format>`, :ref:`TPR <TPR-format>`, :ref:`XML <XML-format>`',
    ]),
    list([
      'impropers',
      'impropers',
      ':ref:`DATA <DATA-format>`, :ref:`GSD <GSD-format>`, :ref:`PSF <PSF-format>`, :ref:`TOP, PRMTOP, PARM7 <TOP-format>`, :ref:`TPR <TPR-format>`, :ref:`XML <XML-format>`',
    ]),
  ])
# ---
# name: test_TopologyAttrs
  list([
    list([
      'altLoc',
      'altLocs',
      'Alternate location',
      ':ref:`MMTF <MMTF-format>`, :ref:`PDB, ENT <PDB-format>`, :ref:`PDBQT <PDBQT-format>`, :ref:`XPDB <XPDB-format>`',
    ]),
    list([
      'angles',
      'angles',
      '',
      ':ref:`DATA <DATA-format>`, :ref:`GSD <GSD-format>`, :ref:`PSF <PSF-format>`, :ref:`TOP, PRMTOP, PARM7 <TOP-format>`, :ref:`TPR <TPR-format>`, :ref:`XML <XML-format>`',
    ]),
    list([
      'aromaticity',
      'aromaticities',
      '',
      '',
    ]),
    list([
      'atomiccharge',
      'atomiccharges',
      'Atomic number',
      ':ref:`GMS <GMS-format>`',
    ]),
    list([
      'atomnum',
      'atomnums',
      '?',
      ':ref:`DMS <DMS-format>`',
    ]),
    list([
      'bonds',
      'bonds',
      '',
      ':ref:`DATA <DATA-format>`, :ref:`DMS <DMS-format>`, :ref:`GSD <GSD-format>`, :ref:`MMTF <MMTF-format>`, :ref:`MOL2 <MOL2-format>`, :ref:`PSF <PSF-format>`, :ref:`TOP, PRMTOP, PARM7 <TOP-format>`, :ref:`TPR <TPR-format>`, :ref:`TXYZ, ARC <TXYZ-format>`, :ref:`XML <XML-format>`',
    ]),
    list([
      'chainID',
      'chainIDs',
      'chain ID',
      ':ref:`DMS <DMS-format>`, :ref:`PDB, ENT <PDB-format>`, :ref:`XPDB <XPDB-format>`',
    ]),
    list([
      'charge',
      'charges',
      'partial atomic charge',
      ':ref:`DATA <DATA-format>`, :ref:`DMS <DMS-format>`, :ref:`GSD <GSD-format>`, :ref:`MMTF <MMTF-format>`, :ref:`MOL2 <MOL2-format>`, :ref:`PDBQT <PDBQT-format>`, :ref:`PQR <PQR-format>`, :ref:`PSF <PSF-format>`, :ref:`TOP, PRMTOP, PARM7 <TOP-format>`, :ref:`TPR <TPR-format>`, :ref:`XML <XML-format>`',
    ]),
    list([
      'chargegroup',
      'chargegroups',
      '',
      '',
    ]),
    list([
      'chirality',
      'chiralities',
      '',
      '',
    ]),
    list([
      'cmaps',
      'cmaps',
      '',
      '',
    ]),
    list([
      'dihedrals',
      'dihedrals',
      '',
      ':ref:`DATA <DATA-format>`, :ref:`GSD <GSD-format>`, :ref:`PSF <PSF-format>`, :ref:`TOP, PRMTOP, PARM7 <TOP-format>`, :ref:`TPR <TPR-format>`, :ref:`XML <XML-format>`',
    ]),
    list([
      'element',
      'elements',
      'atom element',
      ':ref:`IN, FHIAIMS <IN-format>`, :ref:`MOL2 <MOL2-format>`, :ref:`PDB, ENT <PDB-format>`, :ref:`TPR <TPR-format>`, :ref:`TXYZ, ARC <TXYZ-format>`, :ref:`XYZ <XYZ-format>`',
    ]),
    list([
      'epsilon',
      'epsilons',
      '',
      '',
    ]),
    list([
      'epsilon14',
      'epsilon14s',
      '',
      '',
    ]),
    list([
      'formalcharge',
      'formalcharges',
      '',
      '',
    ]),
    list([
      'gbscreen',
      'gbscreens',
      '',
      '',
    ]),
    list([
      'icode',
      'icodes',
      'atom insertion code',
      ':ref:`MMTF <MMTF-format>`, :ref:`PDB, ENT <PDB-format>`, :ref:`PDBQT <PDBQT-format>`, :ref:`PQR <PQR-format>`, :ref:`XPDB <XPDB-format>`',
    ]),
    list([
      'impropers',
      'impropers',
      '',
      ':ref:`DATA <DATA-format>`, :ref:`GSD <GSD-format>`, :ref:`PSF <PSF-format>`, :ref:`TOP, PRMTOP, PARM7 <TOP-format>`, :ref:`TPR <TPR-format>`, :ref:`XML <XML-format>`',
    ]),
    list([
      'model',
      'models',
      'model number (from 0)',
      ':ref:`MMTF <MMTF-format>`',
    ]),
    list([
      'molnum',
      'molnums',
      '[molecules] number (from 0)',
      ':ref:`TPR <TPR-format>`',
    ]),
    list([
      'moltype',
      'moltypes',
      '[moleculetype] name',
      ':ref:`TPR <TPR-format>`',
    ]),
    list([
      'name',
      'names',
      'atom names',
      ':ref:`CONFIG <CONFIG-format>`, :ref:`CRD <CRD-format>`, :ref:`DMS <DMS-format>`, :ref:`GMS <GMS-format>`, :ref:`GRO <GRO-format>`, :ref:`GSD <GSD-format>`, :ref:`HISTORY <HISTORY-format>`, :ref:`IN, FHIAIMS <IN-format>`, :ref:`MMTF <MMTF-format>`, :ref:`MOL2 <MOL2-format>`, :ref:`PDB, ENT <PDB-format>`, :ref:`PDBQT <PDBQT-format>`, :ref:`PQR <PQR-format>`, :ref:`PSF <PSF-format>`, :ref:`TOP, PRMTOP, PARM7 <TOP-format>`, :ref:`TPR <TPR-format>`, :ref:`TXYZ, ARC <TXYZ-format>`, :ref:`XPDB <XPDB-format>`, :ref:`XYZ <XYZ-format>`',
    ]),
    list([
      'nbindex',
      'nbindices',
      '',
      '',
    ]),
    list([
      'occupancy',
      'occupancies',
      'atom occupancy',
      ':ref:`MMTF <MMTF-format>`, :ref:`PDB, ENT <PDB-format>`, :ref:`PDBQT <PDBQT-format>`, :ref:`XPDB <XPDB-format>`',
    ]),
    list([
      'radius',
      'radii',
      'atomic radius',
      ':ref:`GSD <GSD-format>`, :ref:`PQR <PQR-format>`, :ref:`XML <XML-format>`',
    ]),
    list([
      'record_type',
      'record_types',
      'ATOM / HETATM',
      ':ref:`PDB, ENT <PDB-format>`, :ref:`PDBQT <PDBQT-format>`, :ref:`PQR <PQR-format>`, :ref:`XPDB <XPDB-format>`',
    ]),
    list([
      'resname',
      'resnames',
      'residue name (except GSD has ints)',
      ':ref:`CRD <CRD-format>`, :ref:`DMS <DMS-format>`, :ref:`GRO <GRO-format>`, :ref:`GSD <GSD-format>`, :ref:`MMTF <MMTF-format>`, :ref:`MOL2 <MOL2-format>`, :ref:`PDB, ENT <PDB-format>`, :ref:`PDBQT <PDBQT-format>`, :ref:`PQR <PQR-format>`, :ref:`PSF <PSF-format>`, :ref:`TOP, PRMTOP, PARM7 <TOP-format>`, :ref:`TPR <TPR-format>`, :ref:`XPDB <XPDB-format>`',
    ]),
    list([
      'rmin',
      'rmins',
      '',
      '',
    ]),
    list([
      'rmin14',
      'rmin14s',
      '',
      '',
    ]),
    list([
      'solventradius',
      'solventradii',
      '',
      '',
    ]),
    list([
      'tempfactor',
      'tempfactors',
      'B-factor',
      ':ref:`CRD <CRD-format>`, :ref:`MMTF <MMTF-format>`, :ref:`PDB, ENT <PDB-format>`, :ref:`PDBQT <PDBQT-format>`, :ref:`XPDB <XPDB-format>`',
    ]),
    list([
      'type_index',
      'type_indices',
      'amber atom type number',
      ':ref:`TOP, PRMTOP, PARM7 <TOP-format>`',
    ]),
    list([
      'ureybradleys',
      'ureybradleys',
      '',
      '',
    ]),
  ])
# ---
# name: test_TopologyParsers
  list([
    list([
      ':ref:`CONFIG <CONFIG-format>`',
      'DL_Poly CONFIG file',
      'names',
      'masses, types',
    ]),
    list([
      ':ref:`CRD <CRD-format>`',
      'CHARMM CARD file',
      'names, resnames, tempfactors',
      'masses, types',
    ]),
    list([
      ':ref:`DATA <DATA-format>`',
      'LAMMPS data file',
      'angles, bonds, charges, dihedrals, impropers',
      '',
    ]),
    list([
      ':ref:`DMS <DMS-format>`',
      'DESRES Molecular Structure file',
      'atomnums, bonds, chainIDs, charges, names, resnames',
      'types',
    ]),
    list([
      ':ref:`GMS <GMS-format>`',
      'GAMESS file',
      'atomiccharges, names',
      'masses, types',
    ]),
    list([
      ':ref:`GRO <GRO-format>`',
      'GROMACS structure file',
      'names, resnames',
      'masses, types',
    ]),
    list([
      ':ref:`GSD <GSD-format>`',
      'HOOMD GSD file',
      'angles, bonds, charges, dihedrals, impropers, names, radii, resnames',
      '',
    ]),
    list([
      ':ref:`HISTORY <HISTORY-format>`',
      'DL_Poly HISTORY file',
      'names',
      'masses, types',
    ]),
    list([
      ':ref:`IN, FHIAIMS <IN-format>`',
      'FHI-aims input file',
      'elements, names',
      'masses, types',
    ]),
    list([
      ':ref:`LAMMPSDUMP <LAMMPSDUMP-format>`',
      'LAMMPS ascii dump file',
      '',
      'masses',
    ]),
    list([
      ':ref:`MMTF <MMTF-format>`',
      'MMTF file',
      'altLocs, bonds, charges, icodes, models, names, occupancies, resnames, tempfactors',
      'masses',
    ]),
    list([
      ':ref:`MOL2 <MOL2-format>`',
      'Tripos MOL2 file',
      'bonds, charges, elements, names, resnames',
      'masses',
    ]),
    list([
      ':ref:`PDB, ENT <PDB-format>`',
      'Standard PDB file',
      'altLocs, chainIDs, elements, icodes, names, occupancies, record_types, resnames, tempfactors',
      'masses, types',
    ]),
    list([
      ':ref:`PDBQT <PDBQT-format>`',
      'PDBQT file',
      'altLocs, charges, icodes, names, occupancies, record_types, resnames, tempfactors',
      'masses',
    ]),
    list([
      ':ref:`PQR <PQR-format>`',
      'PQR file',
      'charges, icodes, names, radii, record_types, resnames',
      'masses, types',
    ]),
    list([
      ':ref:`PSF <PSF-format>`',
      'CHARMM, NAMD, or XPLOR PSF file',
      'angles, bonds, charges, dihedrals, impropers, names, resnames',
      '',
    ]),
    list([
      ':ref:`TOP, PRMTOP, PARM7 <TOP-format>`',
      'AMBER topology file',
      'angles, bonds, charges, dihedrals, impropers, names, resnames, type_indices',
      '',
    ]),
    list([
      ':ref:`TPR <TPR-format>`',
      'GROMACS run topology file',
      'angles, bonds, charges, dihedrals, elements, impropers, molnums, moltypes, names, resnames',
      '',
    ]),
    list([
      ':ref:`TXYZ, ARC <TXYZ-format>`',
      'Tinker file',
      'bonds, elements, names',
      'masses',
    ]),
    list([
      ':ref:`XML <XML-format>`',
      'HOOMD XML file',
      'angles, bonds, charges, dihedrals, impropers, radii',
      '',
    ]),
    list([
      ':ref:`XPDB <XPDB-format>`',
      'Extended PDB file',
      'altLocs, chainIDs, icodes, names, occupancies, record_types, resnames, tempfactors',
      'masses, types',
    ]),
    list([
      ':ref:`XYZ <XYZ-format>`',
      'XYZ file',
      'elements, names',
      'masses, types',
    ]),
  ])
# ---
